controllers/watcher_handler.go

```diff
@@ -0,0 +1,96 @@
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type newWatcher func(name string, r *DisruptionReconciler, instance *chaosv1beta1.Disruption) (watchers.Watcher, error)
+
+var watcherList = map[string]newWatcher{
+	"chaos-pod": func(name string, r *DisruptionReconciler, instance *chaosv1beta1.Disruption) (watchers.Watcher, error) {
+		return watchers.NewChaosPodWatcher(name, r.Recorder, instance, r.log)
+	},
+	"disruption-target": func(name string, r *DisruptionReconciler, instance *chaosv1beta1.Disruption) (watchers.Watcher, error) {
+		return watchers.NewDisruptionTargetWatcher(name, r.Recorder, r.Reader, r.EnableObserver, instance, r.MetricsSink, r.log)
+	},
+}
+
+// manageInstanceWatchers manage the watchers for each disruption instance
+func (r *DisruptionReconciler) manageInstanceWatchers(instance *chaosv1beta1.Disruption) error {
+	r.log.Debugf("start to manage watchers for the %s disruption", instance.Name)
+
+	if len(watcherList) == 0 {
+		return nil
+	}
+
+	if err := r.WatcherManager.RemoveExpiredWatchers(); err != nil {
+		return err
+	}
+
+	for wName, newWatcherFunc := range watcherList {
+		watcherName, err := getInstanceWatcherName(wName, instance)
+		if err != nil {
+			return fmt.Errorf("failed to initialize %s watcher of the disruption %s: could not get the watcher name. Error: %s", wName, instance.Name, err.Error())
+		}
+
+		if w := r.WatcherManager.GetWatcher(watcherName); w != nil {
+			continue
+		}
+
+		watcherInstance, err := newWatcherFunc(watcherName, r, instance)
+		if err != nil {
+			r.log.Errorf("failed to initialize the %s watcher of the %s disruption. Error: %s", wName, instance.Name, err.Error())
+			return err
+		}
+
+		if err := r.WatcherManager.AddWatcher(watcherInstance); err != nil {
+			r.log.Errorf("could not add the watcher %s of the  %s disruption. Error: %s", wName, instance.Name, err.Error())
+			return err
+		}
+
+		r.log.Debugf("the %s watcher of the %s disruption is created", wName, instance.Name)
+	}
+
+	return nil
+}
+
+// clearInstanceWatchers remove watchers of the disruption
+func (r *DisruptionReconciler) clearInstanceWatchers(instance *chaosv1beta1.Disruption) {
+	r.log.Debugf("start cleaning watchers for the %s disruption", instance.Name)
+
+	for wName := range watcherList {
+		watcherName, err := getInstanceWatcherName(wName, instance)
+		if err != nil {
+			r.log.Errorf("could not remove the %s watcher of the %s disruption: error getting watcher name. Error: %s", wName, instance.Name, err.Error())
+			continue
+		}
+
+		if w := r.WatcherManager.GetWatcher(watcherName); w != nil {
+			if err := r.WatcherManager.RemoveWatcher(w); err != nil {
+				r.log.Errorf("could not remove the %s watcher of the %s disruption: Error: %s", wName, instance.Name, err.Error())
+			}
+
+			r.log.Debugf("the %s watcher of the %s disruption is removed", wName, instance.Name)
+		}
+	}
+}
+
+func getInstanceWatcherName(watcherPrefixName string, instance *chaosv1beta1.Disruption) (string, error) {
```

Comment:
Currently, the `WatcherManager` just manage `Watcher` instances, but it does not have a notion of `Disruption`. The factory should take in argument a disruption and create a watcher so it could handle the naming of the watcher based on the disruption. I agree with you it will be simpler to have a `map[string][]Watcher` inside the WatcherManager, but it should be renamed `DisuptionsWatchersManger` or another name _i am not comfortable to name things..._. I will think about a new implementation, but It will introduce more changes.